File: backend/app/schemas/batch.py
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field

class BatchCreateRequest(BaseModel):
    crop_name: str
    variety: Optional[str] = None
    quantity: float = Field(gt=0, description="Quantity must be greater than 0")
    unit: str = "kg"
    farmer_location: str
    mandi_owner_id: str
    harvest_date: datetime
    notes: Optional[str] = None

class BatchStatusUpdateRequest(BaseModel):
    status: str  # PENDING | ACCEPTED | IN_TRANSIT | RECEIVED | REJECTED
    notes: Optional[str] = None

class BatchResponse(BaseModel):
    id: str
    crop_name: str
    variety: Optional[str] = None
    quantity: float
    unit: str
    farmer_id: str
    farmer_name: str
    farmer_email: str
    farmer_location: str
    mandi_owner_id: str
    mandi_name: str
    mandi_location: Optional[str] = None
    status: str
    harvest_date: datetime
    created_at: datetime
    notes: Optional[str] = None

class BatchPaginatedResponse(BaseModel):
    items: List[BatchResponse]
    total: int
    page: int
    size: int
    pages: int

class MandiOptionResponse(BaseModel):
    id: str
    name: str
    mandi_name: str
    mandi_location: Optional[str] = None

File: frontend/src/api/batch.ts
import { apiClient } from "@/api/http";

export interface Batch {
  id: string;
  crop_name: string;
  variety?: string;
  quantity: number;
  unit: string;
  farmer_id: string;
  farmer_name: string;
  farmer_email: string;
  farmer_location: string;
  mandi_owner_id: string;
  mandi_name: string;
  mandi_location?: string;
  status: "PENDING" | "ACCEPTED" | "IN_TRANSIT" | "RECEIVED" | "REJECTED";
  harvest_date: string;
  created_at: string;
  notes?: string;
}

export interface BatchCreateInput {
  crop_name: string;
  variety?: string;
  quantity: number;
  unit: string;
  farmer_location: string;
  mandi_owner_id: string;
  harvest_date: string;
  notes?: string;
}

export interface BatchPaginatedResponse {
  items: Batch[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface MandiOption {
  id: string;
  name: string;
  mandi_name: string;
  mandi_location?: string;
}

export async function fetchMandiOptions(): Promise<MandiOption[]> {
  const res = await apiClient.get("/api/mandis");
  return res.data;
}

export async function createBatch(data: BatchCreateInput): Promise<Batch> {
  const res = await apiClient.post("/api/batches", data);
  return res.data;
}

export async function fetchBatches(params?: {
  page?: number;
  size?: number;
  search?: string;
  status?: string;
}): Promise<BatchPaginatedResponse> {
  const res = await apiClient.get("/api/batches", { params });
  return res.data;
}

export async function fetchBatchById(id: string): Promise<Batch> {
  const res = await apiClient.get(`/api/batches/${id}`);
  return res.data;
}

export async function updateBatchStatus(
  batchId: string,
  status: string,
  notes?: string
): Promise<Batch> {
  const res = await apiClient.patch(`/api/batches/${batchId}/status`, {
    status,
    notes,
  });
  return res.data;
}
